### PR TITLE
Add IDs to credentials based on their name

### PR DESCRIPTION
**Pull request by AndyTitu**
## Overview

We need a way to identify credentials in a stable long-term manner. The credential names could be subject to change at a certain point in the future, so we are introducing an additional layer meant for unique identification that is not vulnerable to sdk changes.

Also we need to enforce that plugins are uniquely identified by their name, and credentials inside each plugin definition are uniquely identifiable by their ID.

## Type of change
 

- [ ] Created a new plugin
- [ ] Improved an existing plugin
- [ ] Fixed a bug in an existing plugin
- [x] Improved contributor utilities or experience

## Related Issue(s)
  

* Resolves: #257 

Relates other some internal issues.

## Changelog
  

Credentials now have an ID they can be identified after.

Plugins can be uniquely identified by their name, and credentials inside a plugin definition can be uniquely identified by their ID

### CODE REVIEW

File: sdk/schema/credname/names_test.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+		"cli_token",
+		"credential",
+		"credentials",
+		"database_credentials",
+		"login_details",
+		"personal_api_token",
+		"personal_access_token",
+		"registry_credentials",
+		"secret_key",
+	}
```
Comment (florisvdg) [inline sdk/schema/credname/names_test.go:49]: Are we expecting contributors to keep this list up to date? If not, I'd recommend making this slightly less realistic-looking and more like a unit test that is also testing more edge cases.

File: sdk/schema/validation.go [hunk 2]
```diff
@@ -112,3 +112,27 @@ func CredentialReferencesInCredentialList(plugin Plugin) bool {
 	}
 	return true
 }
+
+func CredentialsAreUniquelyIdentifiableInsideAPlugin(plugin Plugin) bool {
```
Comment (florisvdg) [inline sdk/schema/validation.go:116]: How about something like:
```suggestion
func PluginHasNoDuplicateCredentials(plugin Plugin) bool {
```
Or:
```suggestion
func NoDuplicateCredentials(plugin Plugin) bool {
```

Comment (AndyTitu) [reply sdk/schema/validation.go]: Done, thanks!

File: sdk/schema/credname/names_test.go (hunk 1 above)
Comment (AndyTitu) [reply sdk/schema/credname/names_test.go]: Added some additional edge case in addition to the list of the copy pasted existing credential names.

File: sdk/names.go [hunk 3]
```diff
@@ -15,3 +17,19 @@ type CredentialName string
...
+
+func (n CredentialName) ID() CredentialTypeID {
+	return CredentialTypeID(credentialNameToSnakeCase(n))
+}
+
+func credentialNameToSnakeCase(name CredentialName) string {
+	str := name.String()
+	str = strings.ReplaceAll(str, " ", "_")
+	return strings.ToLower(str)
+}
```
[the remark this answers is not in the document]
Comment (AndyTitu) [reply sdk/names.go]: Done, thanks!

[the remark this answers is not in the document]
Comment (AndyTitu) [reply sdk/names.go]: Should we expect commas (,) as well?

[the remark this answers is not in the document]
Comment (florisvdg) [reply sdk/names.go]: `,` and `.` could make sense to add as well 👍

[the remark this answers is not in the document]
Comment (AndyTitu) [reply sdk/names.go]: Done, and added more tests. Thanks

**Review by jpcoenen — APPROVED** [review]

**Review by florisvdg — APPROVED** [review]

File: plugins/registry_test.go [hunk 4]
```diff
@@ -17,3 +18,12 @@ func TestValidatePlugins(t *testing.T) {
...
 }
+
+func TestAllPluginsHaveUniqueNames(t *testing.T) {
+	var pluginNames []string
+	for _, p := range registry {
+		pluginNames = append(pluginNames, p.Name)
+	}
+
+	assert.True(t, schema.IsStringSliceASet(pluginNames))
+}
```
Comment (jpcoenen) [inline plugins/registry_test.go:29]: We could add a similar test that also makes sure the credential ID's used within a plugin are unique.

**Review by hculea — APPROVED** [review]
